p2p/kademlia/dht.go

```diff
@@ -35,661 +39,436 @@ type Options struct {
+				defer cancel()
+
+				// invoke the request and handle the response
+				response, err := s.network.Call(ctx, request)
+				if err != nil {
+					log.WithContext(ctx).Errorf("network call: %v", err)
+					return
 				}
-			})
-		}
+				log.WithContext(ctx).Debugf("ping response: %v", response.String())
 
-		if err := group.Wait(); err != nil {
-			return err
+				// add the node to the route table
+				s.addNode(response.Sender)
+			}()
 		}
 	}
 
-	if dht.NumNodes() > 0 {
-		_, _, err := dht.iterate(ctx, iterateFindNode, dht.ht.Self.ID, nil)
-		return err
+	// wait until all are done
+	wg.Wait()
+
+	// if it has nodes in local route tables
+	if s.ht.totalCount() > 0 {
+		// iterative find node from the nodes
+		if _, err := s.iterate(ctx, IterateFindNode, s.ht.self.ID, nil); err != nil {
+			log.WithContext(ctx).Errorf("iterative find node: %v", err)
+			return err
+		}
 	}
 
 	return nil
 }
 
-// Disconnect will trigger a disconnect from the network. All underlying sockets
-// will be closed.
-func (dht *DHT) Disconnect() error {
-	// TODO if .CreateSocket() is called, but .Listen() is never called, we
-	// don't provide a way to close the socket
-	return dht.networking.disconnect()
+// new a message
+func (s *DHT) newMessage(messageType int, receiver *Node, data interface{}) *Message {
+	return &Message{
+		Sender:      s.ht.self,
+		Receiver:    receiver,
+		MessageType: messageType,
+		Data:        data,
+	}
 }
 
-// Iterate does an iterative search through the network. This can be done
-// for multiple reasons. These reasons include:
-//     iterativeStore - Used to store new information in the network.
-//     iterativeFindNode - Used to bootstrap the network.
-//     iterativeFindValue - Used to find a value among the network given a key.
-func (dht *DHT) iterate(ctx context.Context, t int, target []byte, data []byte) (value []byte, closest []*NetworkNode, err error) {
-	sl := dht.ht.getClosestContacts(alpha, target, []*NetworkNode{})
-
-	// We keep track of nodes contacted so far. We don't contact the same node
-	// twice.
-	var contacted = make(map[string]bool)
+func (s *DHT) doMultiWorkers(ctx context.Context, iterativeType int, target []byte, nl *NodeList, contacted map[string]bool, haveRest bool) []*Message {
+	var number int
+
+	var mux sync.Mutex
+	// nodes which will be removed
+	removedNodes := []*Node{}
+	// responses from remote node
+	responses := []*Message{}
+
+	var wg sync.WaitGroup
+	// send the message to the first (closest) alpha nodes
+	for _, node := range nl.Nodes {
+		// only contact alpha nodes
+		if number >= Alpha && !haveRest {
+			break
+		}
+		// ignore the contacted node
+		if contacted[string(node.ID)] {
+			continue
+		}
+		contacted[string(node.ID)] = true
+
+		// update the running goroutines
+		number++
+
+		log.WithContext(ctx).Debugf("start work %v for node(%d): %s", iterativeType, len(nl.Nodes), node.String())
+
+		wg.Add(1)
+		// send and recive message concurrently
```

Comment:
seem that we have to wait all responses from all peers before do processing responses.
So could we can change it to like:
```
func (s *DHT) doMultiWorkers(ctx,...) chan *Message {
	responseChan := make(chan * Message)
	go {
		var wg sync.WaitGroup
		// send the message to the first (closest) alpha nodes
		for _, node := range nl.Nodes {
			....
			wg.Add(1)
			// send and recive message concurrently
			go func(receiver *Node) {
				defer wg.Done()
				...
				...
				responseChan <- newMessage 
			}
		}

		wg.Wait()
		close(responseChan)
	}
}


for message := doMultiWorkers(ctx,...) {
	
}
```

Reply:
according to the whole logic, it's few different between the two ways, but i like the current way, i think it's more clean

Reply:
That is different : for your one : you wait for all responses available before processing.
I suggest handle any response when it's available. That can have better performance, like in case poor connectivity

Reply:
done